### PR TITLE
Upgrade to Ktor 2

### PR DESCRIPTION
**Pull request by DRSchlaubi**
this is a WIP PR to make upgrading to Ktor 2 easier once it becomes stable

### CODE REVIEW

**Comment by BenWoodworth** [comment]
Just a heads up, Ktor 2.0's release might be just a few days away. They're [waiting on Kotlin 1.6.20](https://blog.jetbrains.com/ktor/2022/02/11/ktor-roadmap-what-s-next/), and [1.6.20 is being staged for release](https://plugins.jetbrains.com/plugin/6954-kotlin/versions/staging) :)

File: gateway/src/main/kotlin/DefaultGateway.kt [hunk 1]
```diff
@@ -159,7 +159,7 @@ public class DefaultGateway(private val data: DefaultGatewayData) : Gateway {
 
 
     private suspend fun readSocket() {
-        socket.incoming.asFlow().buffer(Channel.UNLIMITED).collect {
+        socket.incoming.asFlow().buffer(Channel.UNLIMITED).collectLatest {
```
Comment (HopeBaron) [inline gateway/src/main/kotlin/DefaultGateway.kt:162]: I remember this was due to a bug in previous version of coroutine lib, should this be changed back?

Comment (Lukellmann) [reply gateway/src/main/kotlin/DefaultGateway.kt:162]: yes, it should, `collectLatest` does a different thing

Comment (DRSchlaubi) [reply gateway/src/main/kotlin/DefaultGateway.kt:162]: I remember they marked this as internal and then not made it internal again

File: gateway/src/samples/kotlin/EventListener.kt [hunk 2]
```diff
@@ -19,8 +18,8 @@ suspend fun main(args: Array<String>) {
     val gateway = DefaultGateway {
         client = HttpClient(CIO) {
             install(WebSockets)
-            install(JsonFeature) {
-                serializer = KotlinxSerializer(Json)
+            install(ContentNegotiation) {
+                json()
```
Comment (Lukellmann) [inline gateway/src/samples/kotlin/EventListener.kt:22]: maybe here too, I'm not sure whether the previously passed `Json` was also custom

File: gateway/src/test/kotlin/gateway/DefaultGatewayTest.kt [hunk 3]
```diff
@@ -31,8 +33,8 @@ class DefaultGatewayTest {
         val gateway = DefaultGateway {
             client = HttpClient(CIO) {
                 install(WebSockets)
-                install(JsonFeature) {
-                    serializer = KotlinxSerializer(Json)
+                install(ContentNegotiation) {
+                    json()
```
Comment (Lukellmann) [inline gateway/src/test/kotlin/gateway/DefaultGatewayTest.kt:37]: and here?

File: rest/src/main/kotlin/Image.kt [hunk 4]
```diff
@@ -22,11 +22,12 @@ public class Image private constructor(public val data: ByteArray, public val fo
...
 
         public suspend fun fromUrl(client: HttpClient, url: String): Image = with(Dispatchers.IO) {
-            val call = client.request<HttpResponse>(url) { method = HttpMethod.Get }
+            val call = client.request(url) { method = HttpMethod.Get }
             val contentType = call.headers["Content-Type"]
                 ?: error("expected 'Content-Type' header in image request")
 
-            val bytes = call.content.toByteArray()
+            @Suppress("EXPERIMENTAL_API_USAGE")
+            val bytes = call.bodyAsChannel().toByteArray()
```
Comment (Lukellmann) [inline rest/src/main/kotlin/Image.kt:30]: I saw `call.body<ByteArray>()` used in the migration guide instead: https://ktor.io/docs/migrating-2.html#responses

File: rest/src/main/kotlin/Image.kt [hunk 5]
```diff
@@ -22,11 +22,12 @@ public class Image private constructor(public val data: ByteArray, public val fo
         }
 
         public suspend fun fromUrl(client: HttpClient, url: String): Image = with(Dispatchers.IO) {
-            val call = client.request<HttpResponse>(url) { method = HttpMethod.Get }
+            val call = client.request(url) { method = HttpMethod.Get }
```
Comment (Lukellmann) [inline rest/src/main/kotlin/Image.kt:25]: can't this simply use `client.get()`?

**Review by Lukellmann — CHANGES_REQUESTED** [review]

File: voice/src/main/kotlin/gateway/DefaultVoiceGateway.kt [hunk 6]
```diff
@@ -130,7 +130,7 @@ public class DefaultVoiceGateway(
     }
 
     private suspend fun readSocket() {
-        socket.incoming.asFlow().buffer(Channel.UNLIMITED).collect {
+        socket.incoming.asFlow().buffer(Channel.UNLIMITED).collectLatest {
```
Comment (Lukellmann) [inline voice/src/main/kotlin/gateway/DefaultVoiceGateway.kt:133]: also replace here

**Review by Lukellmann — CHANGES_REQUESTED** [review]

File: gateway/src/samples/kotlin/EventListener.kt (hunk 2 above)
Comment (DRSchlaubi) [reply gateway/src/samples/kotlin/EventListener.kt:22]: It just passed in `Json.Companion`, which also was the default back then

File: gateway/src/test/kotlin/gateway/DefaultGatewayTest.kt (hunk 3 above)
Comment (DRSchlaubi) [reply gateway/src/test/kotlin/gateway/DefaultGatewayTest.kt:37]: It just passed in `Json.Companion`, which also was the default back then

File: gateway/src/samples/kotlin/EventListener.kt (hunk 2 above)
Comment (Lukellmann) [reply gateway/src/samples/kotlin/EventListener.kt:22]: then we don't need it here 👍

File: rest/src/main/kotlin/Image.kt [hunk 7]
```diff
@@ -22,11 +20,12 @@ public class Image private constructor(public val data: ByteArray, public val fo
...
 
         public suspend fun fromUrl(client: HttpClient, url: String): Image = with(Dispatchers.IO) {
-            val call = client.request<HttpResponse>(url) { method = HttpMethod.Get }
+            val call = client.get(url)
             val contentType = call.headers["Content-Type"]
                 ?: error("expected 'Content-Type' header in image request")
 
-            val bytes = call.content.toByteArray()
+            @Suppress("EXPERIMENTAL_API_USAGE")
+            val bytes = call.body<ByteArray>()
```
Comment (Lukellmann) [inline rest/src/main/kotlin/Image.kt:28]:
```suggestion
            val bytes = call.body<ByteArray>()
```
not necessary, only the `content` property used previously became internal api in ktor 2

File: core/src/test/kotlin/regression/CacheMissRegression.kt [hunk 8]
```diff
@@ -83,7 +81,7 @@ object FakeGateway : Gateway {
 class CrashingHandler(val client: HttpClient, override val token: String) : RequestHandler {
     override suspend fun <B : Any, R> handle(request: Request<B, R>): R {
         if (request.route != Route.CurrentUserGet) throw IllegalStateException("shouldn't do a request")
-        val response = client.request<HttpStatement> {
+        val response = client.prepareRequest {
```
Comment (Lukellmann) [inline core/src/test/kotlin/regression/CacheMissRegression.kt:84]: `prepareRequest { ... }.execute()` can be replaced with just `request { ... }`

**Review by Lukellmann — APPROVED** [review]

**Pull request merged**